src/Funnel.sol

```diff
@@ -198,23 +198,36 @@ contract Funnel is IFunnel, NativeMetaTransaction, MetaTxContext, Initializable
     /// @notice fetch amounts spendable by _spender
     /// @return remaining allowance at the current point in time
     function allowance(address _owner, address _spender)
-        public
+        external
         view
         returns (uint256 remaining)
     {
         return _remainingAllowance(_owner, _spender);
     }
 
+    // @notice fetch remaining allowance between _owner and _spender while accounting for base token allowance.
     function _remainingAllowance(address _owner, address _spender)
         private
         view
         returns (uint256)
     {
         RenewableAllowance memory a = rAllowance[_owner][_spender];
+        uint256 baseAllowance = _baseToken.allowance(_owner, address(this));
+        uint256 recovered = a.recoveryRate * uint64(block.timestamp - a.lastUpdated); // uint192 * uint64 = uint256
+        uint256 remainingAllowance = saturatingAdd(a.remaining, recovered); // uint256 + uint256 potential overflow
+        uint256 currentAllowance = remainingAllowance > a.maxAmount
+            ? a.maxAmount
+            : remainingAllowance;
+        return currentAllowance > baseAllowance ? baseAllowance : currentAllowance;
+    }
 
-        uint256 recovered = a.recoveryRate * (block.timestamp - a.lastUpdated);
-        uint256 remainingAllowance = a.remaining + recovered;
-        return remainingAllowance > a.maxAmount ? a.maxAmount : remainingAllowance;
+    // @dev returns the sum of two uint256 values, saturating at 2**256 - 1
+    function saturatingAdd(uint256 a, uint256 b) internal pure returns (uint256) {
+        unchecked {
+            uint256 c = a + b;
+            if (c < a) return type(uint256).max;
```

Comment:
Why do we need to - 1?

Reply:
```
function transferFrom(
    address from,
    address to,
    uint256 amount
) public returns (bool) {
    uint256 remainingAllowance = _remainingAllowance(from, _msgSender());
    if (remainingAllowance < amount) {
        revert InsufficientRenewableAllowance({ available: remainingAllowance });
    }
    if (remainingAllowance != type(uint256).max) {
        rAllowance[from][_msgSender()].remaining = remainingAllowance - amount;
        rAllowance[from][_msgSender()].lastUpdated = uint64(block.timestamp);
    }

(...)
```

to avoid `renewableAllowance` being equal to `type(uint256).max` when not really intended, since that's a special value that indicates that the allowance is to be infinite

by capping it at `type(uint256).max - 1`, we make sure that `transferFrom()` will still deduct from allowance when the previous allowance + recovered amount exceeds `type(uint256).max`

e.g.:
- user with maxAmount 2**256-2 (uint256 max - 1) and recovery rate 1
- if `transferFrom(user, user2, 2**256-2)` is called immediately after approval (eg in same block) then user's allowance will be 0
- if we call it 1 second later then user's allowance will remain at uint256 max

correct me if I'm wrong, but it seems like that is what will happen

Reply:
Yes, it will, so recovering to max uint will cause it to not deduct allowance. 

I think the `-1` should happen not within the add function while making sure recovery can only reach uint256-1.

Reply:
An alternative would be to skip allowance deduction only when `a.maxAmount == type(uint256).max`

Reply:
`type(uint256).max - 1` would also stop transfer of max uint256 when we give renewableApproval of initial: uint256max, with recovery > 0.

Reply:
Moved this to an issue #60 

However for deeper exploration, this issue likely won't happen, see comment in #60